pkg/resources/gce/gce.go

```diff
@@ -797,24 +799,70 @@ func (d *clusterDiscoveryGCE) matchesClusterNameMultipart(name string, maxParts
 }
 
 func (d *clusterDiscoveryGCE) listGCEDNSZone() ([]*resources.Resource, error) {
-	// We never delete the hosted zone, because it is usually shared and we don't create it
-	return nil, nil
-	// TODO: When shared resource PR lands, reintroduce
-	//if dns.IsGossipHostname(d.clusterName) {
-	//	return nil, nil
-	//}
-	//zone, err := d.findDNSZone()
-	//if err != nil {
-	//	return nil, err
-	//}
-	//
-	//return []*resources.Resource{
-	//	{
-	//		Name:    zone.Name(),
-	//		ID:      zone.Name(),
-	//		Type:    "DNS Zone",
-	//		Deleter: d.deleteDNSZone,
-	//		Obj:     zone,
-	//	},
-	//}, nil
+
+	var zones []*dns.ManagedZone
+	var resourceTrackers []*resources.Resource
+
+	zoneResponse, err := d.gceCloud.NameService().ManagedZones.List(d.gceCloud.Project()).Do()
+	if err != nil {
+		return nil, fmt.Errorf("error getting GCE DNS zones %v", err)
+	}
+	for _, zone := range zoneResponse.ManagedZones {
+		zoneName := "." + strings.TrimSuffix(zone.DnsName, ".")
+		if strings.HasSuffix(d.clusterName, zoneName) {
+			zones = append(zones, zone)
+		}
+	}
+
+	for _, zone := range zones {
+		response, err := d.gceCloud.NameService().ResourceRecordSets.List(d.gceCloud.Project(), zone.Name).Do()
+		if err != nil {
+			return nil, fmt.Errorf("error getting GCE DNS zone data %v", err)
+		}
+		for _, record := range response.Rrsets {
+			// adapted from AWS implementation
+			if record.Type != "A" {
+				continue
+			}
+
+			if !strings.HasSuffix(record.Name, "."+d.clusterName+".") {
```

Comment:
Regardless, the logic feels a bit arbitrary to me. If I'm reading this correctly, you're basically trying to figure out if Cloud DNS records are records that we care about (and thus would need to delete). Would it be useful to pull out the logic to figure out if a given record is one that we should be cleaning up after into a helper? 
 
One major thing that I think we're not yet considering, but was eluded to in the comments of the faux-original implementation (commented out) is the concept of "shared" ownership; ie: assuming there is a record in cloud dns that we think we/kops should own, do we really own it? In this case, i'm not trying to poke holes in your implementation, i'm trying to figure out how to clean up after ourselves... I don't have an answer or suggestion right now.

Reply:
This is taken from the AWS implementation and it will just check for api. api.internal.clusterDNSName, api.clusterDNSName, bastion.clusterDNSName. 

I have found a buildPrecreateDNSHostnames in dns.go (but feature flag gated further up) which at first glance could deliver all possible host names. 

For shared records: My assumption was that kops manages its own DNS records (not zones) and they are unique over kops clusters since they include the cluster name as dns zone. So the records should be owned by kops. 

For conflicts: there is the possibility to have a public and private zone with the same name (e.g. example.com). Kops would require a DNSZone as identifier (from code inspection), which the code doesn't honor.

Another conflict: having a zone example.com and cluster.example.com in CloudDNS: They can both contain the record api.cluster.example.com. Tried that out with cloud DNS and no error. But IMHO a misconfiguration in the first place.

Reply:
> This is not to say that the code may not delete a record e.g. in a private only zone with the same DNS name as the public zone. But from my tests I cannot create a cluster where there is a private a public zone with the same DNS name.

I think we can probably not worry about shared ownership of Cloud DNS records, but I think it's possible to end up in a place where you have a shared Zone. Whether or not kops supports that right now for gce is an open question.

For zones, would it be possible to add a label like we have in many AWS resources? I was thinking that if a zone was specified that already existed, we could add label: `"kubernetes.io/cluster/" + clusterName = "shared"`. Then we could skip deletion of that zone. 

This seems like a great place for some tests to make sure that we're cleaning up when we need to and not deleting shared resources, but I'm not sure which route to follow without doing some more thinking. @rifelpet Do you have any suggestions?

Reply:
Using labels sounds optimal. Indeed for AWS resources we dont want kops to manage, kops supports detecting a shared label/tag and if it is present, it wont delete the resource. In AWS it is mostly around networking (VPC, subnets, etc) but I think it makes sense to use it for DNS zones as well.